src/librustc/ty/layout.rs

```diff
@@ -546,7 +546,8 @@ pub struct Struct {
 
     pub min_size: Size,
 
-    /// Size of padding needed by the struct due to requested repr alignment
+    /// Size of padding needed by the struct due to requested repr alignment.
+    /// This is created as required and will be empty if the structs fields require no padding.
     pub padding: Vec<Size>,
```

Comment:
I'm tempted to say that we should have offsets and field sizes, which is more general, and then creating a LLVM type can compute how much padding to use between each two consecutive fields. Check out https://github.com/solson/miri/pull/146#issuecomment-282305202, where I have a potential use for this: removing the implied discriminant field of `enum` variant `Struct`s, which in an `Vec<(field_offset, field_size)>` encoding just be the first field starting e.g. `4` bytes in and thus the LLVM type having a padding field there.

We might also be able to *always* emit padding fields (sometimes zero-sized) and get LLVM to play along.
cc @rust-lang/compiler Anyone think this is terrible idea? IMO it might be better than `memory_index_with_padding_fields` which makes field access `O(n)` instead of `O(1)`.